File: bbenchmarks/americanPutOption.R

{
library("compiler")
enableJIT(3)

M <- as.integer(commandArgs(TRUE)[[2]])
N <- as.integer(sqrt(as.integer(commandArgs(TRUE)[[1]])/M))

americanPut <- function(T, S, K, r, sigma, q, n) {
	deltaT <- T/n
	up <- exp((sigma * sqrt(deltaT)))
	p0 <- (up * exp(-r * deltaT) - exp(-q*deltaT)) * up / (up^2 - 1)
	p1 <- exp(-r * deltaT) - p0

    p <- list()

    i <- 1
    while(i <= n) {
        p[[i]] <- pmax(0, K - S * up^(2*i-n))
	    i <- i+1
    }

    j <- n-1L
    while(j >= 1L) {
        i <- 1L
        while(i <= j) {
            p[[i]] <- pmax( K - S * up^(2*i-j),
                             p0 * p[[i]] + p1 * p[[i+1]] )
		    i <- i+1L
        }
        j <- j-1L
    }
    p[[1]]
}

#cat(americanPut(100, 1:M, 80.0, 3, 2.0, 3.0, N))
cat(system.time(americanPut(100, 1:M, 80.0, 3, 2.0, 3.0, N))[[3]])

}

File: bbenchmarks/americanPutOption.cpp

#include <math.h>
#include <stdint.h>
#include <sys/time.h>
#include <stdlib.h>

#include "../src/common.h"

double* americanPut(double T, double* S, double K, double r, double sigma, double q, int64_t M, int64_t N) {

    double deltaT = T/N;
    double up = exp(sigma * sqrt(deltaT));
    double p0 = (up * exp(-r*deltaT) - exp(-q*deltaT)) * up / (up*up-1);
    double p1 = exp(-r * deltaT) - p0;

    double* p = new double[M*N];
    double* rr = new double[M];


    for(int i = 0; i < N; i++) {
        double pp = pow(up, 2*(i+1)-N);
        for(int64_t j = 0; j < M; j++) {
            p[i*M+j] = std::max(0.0, K-S[j] * pp);
        }
    }

    
    for(int k = N-2; k >= 0; k--) {
        for(int i = 0; i <= k; i++) {
            double pp = pow(up, 2*(i+1)-(k+1));
            for(int64_t j = 0; j < M; j++) {
                p[i*M+j] = std::max( K-S[j] * pp,
                                     p0 * p[i*M+j] + p1 * p[(i+1)*M+j] );
            }
        }
    }

    for(int64_t j = 0; j < M; j++) {
        rr[j] = p[j];
    }

    return rr;
}

int main(int argc, char** argv) {
    static const int64_t M = MM;
    static const int64_t N = NN;

    double* S = new double[M];
    for(int i = 0; i < M; i++)
        S[i] = i+1;

    timespec b = get_time();
    volatile double* r = americanPut(100, S, 80, 3, 2, 3, M, ((int)sqrt(N/M)));
    printf("%f", time_elapsed(b));
}
